servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/RoundRobinSelector.java

```diff
@@ -78,6 +104,92 @@ protected Single<C> selectConnection0(
-        return new RoundRobinSelector<>(index, hosts, getTargetResource(), failOpen);
+        return new RoundRobinSelector<>(index, hosts, getTargetResource(), failOpen, ignoreWeights);
+    }
+
+    private static Scheduler buildScheduler(AtomicInteger index, List<? extends Host<?, ?>> hosts) {
+        boolean allEqualWeights = true;
+        double maxWeight = 0;
+        for (Host<?, ?> host : hosts) {
+            double hostWeight = host.weight();
+            maxWeight = Math.max(maxWeight, hostWeight);
+            allEqualWeights = allEqualWeights && approxEqual(hosts.get(0).weight(), hostWeight);
+        }
+
+        if (allEqualWeights) {
+            return new ConstantScheduler(index, hosts.size());
+        } else {
+            double scaleFactor = MAX_WEIGHT / maxWeight;
+            int[] scaledWeights = new int[hosts.size()];
+
+            for (int i = 0; i < scaledWeights.length; i++) {
+                // Using ceil ensures a few things:
+                // - our max weighted element is picked on every round
+                // - hosts with weights near zero will never be truly zero
+                // - true zero weights will still be truly zero
+                scaledWeights[i] = Math.min(MAX_WEIGHT, (int) Math.ceil(hosts.get(i).weight() * scaleFactor));
+            }
+            return new StrideScheduler(index, scaledWeights);
+        }
+    }
+
+    private abstract static class Scheduler {
+        abstract int nextHost();
+    }
+
+    private static final class ConstantScheduler extends Scheduler {
+
+        private final AtomicInteger index;
+        private final int hostsSize;
+
+        ConstantScheduler(AtomicInteger index, int hostsSize) {
+            this.index = index;
+            this.hostsSize = hostsSize;
+        }
+
+        @Override
+        int nextHost() {
+            return (int) (Integer.toUnsignedLong(index.getAndIncrement()) % hostsSize);
+        }
+    }
+
+    // The stride scheduler is heavily inspired by the Google gRPC implementations with some minor modifications.
+    // The stride scheduler algorithm is convenient in that  it doesn't require synchronization like a priority queue
+    // based solution would which fits well with highly concurrent nature of our HostSelector abstraction.
+    // See the java-grpc library for more details:
+    // https://github.com/grpc/grpc-java/blob/da619e2b/xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java
+    private static final class StrideScheduler extends Scheduler {
+
+        private final AtomicInteger index;
+        private final int[] weights;
+
+        StrideScheduler(AtomicInteger index, int[] weights) {
+            this.index = index;
+            this.weights = weights;
+        }
+
+        @Override
+        int nextHost() {
+            while (true) {
+                long counter = Integer.toUnsignedLong(index.getAndIncrement());
```

Comment:
I would propose to use the same trick here to avoid `long` completely

Reply:
Same as above: I think it's mostly subjective but I like the expanded range without switching to an `AtomicLong`.

Reply:
My expectation is that we can stay in `int` without even going to `long` or `AtomicLong`